bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java

```diff
@@ -0,0 +1,163 @@
+        channelOptionsMap.put(channelUID, options);
+    }
+
+    @Override
+    public @Nullable StateDescription getStateDescription(Channel channel, @Nullable StateDescription original,
+            @Nullable Locale locale) {
+        // can be overridden by subclasses
+        ChannelUID channelUID = channel.getUID();
+        String pattern = channelPatternMap.get(channelUID);
+        List<StateOption> options = channelOptionsMap.get(channelUID);
+        if (pattern == null && options == null) {
+            return null;
+        }
+
+        StateDescriptionFragmentBuilder builder = (original == null) ? StateDescriptionFragmentBuilder.create()
+                : StateDescriptionFragmentBuilder.create(original);
+
+        if (pattern != null) {
+            String localizedPattern = localizeStatePattern(pattern, channel, locale);
+            if (localizedPattern != null) {
+                builder.withPattern(localizedPattern);
+            }
+        }
+
+        if (options != null) {
+            builder.withOptions(localizedStateOptions(options, channel, locale));
+        }
+
+        return builder.build().toStateDescription();
+    }
+
+    /**
+     * Localizes a pattern that should be used for the channel.
+     *
+     * @param pattern a pattern
+     * @param channel the channel
+     * @param locale a locale
+     * @return the localized pattern
+     */
+    protected @Nullable String localizeStatePattern(String pattern, Channel channel, @Nullable Locale locale) {
+        // can be overridden by subclasses
+        ChannelTypeUID channelTypeUID = channel.getChannelTypeUID();
+        if (channelTypeUID != null) {
+            return channelTypeI18nLocalizationService.createLocalizedStatePattern(bundleContext.getBundle(), pattern,
+                    channelTypeUID, locale);
+        }
+        return pattern;
+    }
+
+    /**
+     * Localizes a {@link List} of {@link StateOption}s that should be used for the channel.
+     *
+     * @param options a {@link List} of {@link StateOption}s
+     * @param channel the channel
+     * @param locale a locale
+     * @return the localized {@link List} of {@link StateOption}s
+     */
+    protected List<StateOption> localizedStateOptions(List<StateOption> options, Channel channel,
+            @Nullable Locale locale) {
+        // can be overridden by subclasses
+        ChannelTypeUID channelTypeUID = channel.getChannelTypeUID();
+        if (channelTypeUID != null) {
+            return channelTypeI18nLocalizationService.createLocalizedStateOptions(bundleContext.getBundle(), options,
+                    channelTypeUID, locale);
+        }
+        return options;
+    }
+
+    @Activate
+    protected void activate(ComponentContext componentContext) {
+        bundleContext = componentContext.getBundleContext();
+    }
+
+    @Deactivate
+    public void deactivate() {
+        channelOptionsMap.clear();
+        bundleContext = null;
+    }
+
+    @Reference
```

Comment:
Do not depend on DS annotation inheritance (see e.g. https://community.liferay.com/blogs/-/blogs/bnd-instruction-to-avoid -- it has already been discussion on ESH some time).
If a component use this class a parent, he should define the annotated methods itself and call the parent class implementation.

Reply:
It somehow makes sense. A subclass has to add a `@Component` annotation too to provide its service. I will update the class accordingly but will remove the setter and getter completely. Similar to the `AbstractDiscoveryService` - I always was wondering why it holds instances of two services without requiring them. That is clear now.